### ISSUE TITLE
Build times increased significantly since upgrade to 2.5.6

### ISSUE THREAD

**Issue by icellan** [meteor/meteor#11905]
Build times have exploded and CI is failing after upgrading from 2.5.3 to 2.5.6. At the advice from the community, I upgraded babel to the latest version (v7.17.2) and although this helped, the build times are still very long.

At the request of @renanccastro I'm adding the full profile of the build of our app.

[meteor-profile.txt](https://github.com/meteor/meteor/files/8041552/meteor-profile.txt)

**Comment by darrug** [comment]
I have comparable building time using version 2.5.3 and version 2.6 modified with @babel/parser 7.17.0. Have you tried to update  the package in your local meteor installation? 

Eg.
`.meteor/packages/meteor-tool/.2.6.0.y6uwr.etbh6++os.osx.x86_64+web.browser+web.browser.legacy+web.cordova/mt-os.osx.x86_64/dev_bundle/lib/node_modules/@babel/parser`

**Comment by icellan** [comment]
@darrug I am also on @babel/parser 7.17.0, but build times are very long (the last one was 200 seconds). Just to clarify, it is not only on my OSX laptop but also in CircleCI.

**Comment by MaxTwentythree** [comment]
I'm facing the same issue upgrading from 2.5.5 to 2.5.6 but with build times up to 20 Minutes!
Before it was constantly at a few minutes.
The interesting thing: the build times vary greatly. Sometimes it's a few minutes and sometimes I wait forever. I have no clue what could influence this.

**Comment by afrokick** [comment]
@icellan Could you please provide build log from 2.5.3 version? It is necessary to compare with 2.5.6

**Comment by icellan** [comment]
@afrokick output from 2.5.3.
[meteor-profile-2.5.3.txt](https://github.com/meteor/meteor/files/8049817/meteor-profile-2.5.3.txt)

**Comment by afrokick** [comment]
@icellan Thanks!

@zodern @filipenevola @renanccastro Could you please explain this:

2.5.3:
`ImportScanner#findImportedModuleIdentifiers 8,436 ms (7667)`

2.5.6
` ImportScanner#findImportedModuleIdentifiers 70,183 ms (7563)`

I only found the difference in `reify` vs `@meteorjs/reify` in `js-analyze.js`...

**Comment by make-github-pseudonymous-again** [comment]
@renanccastro Thanks? How do you enable the cache?

**Comment by renanccastro** [comment]
> @renanccastro Thanks? How do you enable the cache?

Caching is automatic. It's used after the first fresh build.

Have anyone taken a shot at the 2.5.7 beta? Is it any better?

**Comment by MaxTwentythree** [comment]
@[renanccastro](https://github.com/renanccastro) I'm trying out 2.5.7-beta.0 and so far it looks good / starts fast again!
I will continue to test it, since the slowdown did not occur at every startup, and report back.
Thanks for the quick support!

**Comment by icellan** [comment]
@renanccastro It is better, but still not as good as before. I just ran 2 builds after each other, before and after upgrading to 2.5.7.-beta.0, and the build time ("Build App") went from 85s to 60s.

**Comment by renanccastro** [comment]
@icellan have you tried to wipe the build cache with meteor reset? or just remove the .meteor/local folder, so we can test with a fresh build/rebuilds.

**Comment by dfuchs2512** [comment]
Can you please merge the fix to the version 2.6?

Thx

**Comment by darrug** [comment]
@renanccastro with 2.5.7-beta.0 I have the same build times as before

**Comment by renanccastro** [comment]
> @renanccastro with 2.5.7-beta.0 I have the same build times as before

You mean the good build times, right? Just to be sure :smile:

**Comment by renanccastro** [comment]
@dfuchs2512 I've published meteor@2.6.1-beta.4 with only this fix. It **does not include** all the changes we have in this 2.6.1 [branch](https://github.com/meteor/meteor/pull/11886). In this way, you can test your app with this version without worrying about other changes.

If this fix works, we will go with 2.6.1 for this fix, and the current 2.6.1 branch will turn 2.6.2.

Please let me know your results.

**Comment by dfuchs2512** [comment]
@renanccastro Just for your information. After update to Version 2.6.1-beta.4 I have the build times back on course like before with 2.5.4. Much better than with 2.6. You fix is working for me.

Thx.

**Comment by lynchem** [comment]
I tried it out and I can confirm times go back to what they were at 2.5.3. We were at 5:49 & 1:40 with 2.5.6 for building after clearing .meteor/local (except db) and building without clearing respectively. Both 2.5.3 & 2.5.7_beta0 clock in at roughly 2:29 & 0:38 
👍🏻

**Comment by MaxTwentythree** [comment]
Same for me! I tested with 2.5.7.-beta.0 now for a day and had no more slow builds!
Thank you for the quick fix!

**Comment by filipenevola** [comment]
> HMR is slow on [2.6.1-beta.4] . anyone having this problem?

Please try 2.7-beta.2.

We have improvements in react-fast-refresh there, not focusing on performance tho

**Comment by ouya99** [comment]
@filipenevola  I tried 2.7-beta.2, but build times are very very slow, compared to 2.6.1-beta4

**Comment by filipenevola** [comment]
Hi, @ouya99 I'm not noticing this.

Could you open a new issue and try to provide a reproduction or at least what is your project setup, technologies, etc

**Comment by ouya99** [comment]
Hi @filipenevola 

what kind of files should I add to the issue? Unfortunately i cannot provide a reproduction. 
HMR currently takes around 50 seconds on each reload.

**Comment by zodern** [comment]
@ouya99 a good first step would be to run meteor with the env var `METEOR_PROFILE=100`, modifying a file used on the client so there is an HMR update, and sharing the logs with us. This will show us what part of the build process is slow.

Also, has HMR always been slow for your app, or is there a specific release that is much faster?

**Comment by ouya99** [comment]
since updating to meteor 2.5.6 and/or meteor 2.6 the HMR is slow. before it was 1-2 seconds update

@filipenevola  attached the log file for a startup and a single file change which resulted in long HMR build
[hmr.txt](https://github.com/meteor/meteor/files/8083136/hmr.txt)

**Comment by zodern** [comment]
Thanks @ouya99. There are two things that are interesting in the profile:
- It seems the cache for findImportedModuleIdentifiers isn't working for your app. With the number of files it is working with, it is possible the LRU cache is too small.
- A lot of time is spent by toStringWithSourceMap. There are many ways we could optimize this, but I've been waiting to see if it actually causes performance issues before doing the work. Now we know it does.

**Comment by ouya99** [comment]
Thanks @filipenevola. Could i try to tweak the LRU cache and how?

**Comment by zodern** [comment]
@ouya99 it would be helpful if you created a new issue for the build performance of your app.

**Comment by znewsham** [comment]
Is there any way to back port this to 2.5? 2.6 is a huge extra release, we've just finished upgrading to 2.5.6 and can't currently move to 2.6 - I think I see in this thread there is a reference to 2.5.7?

FWIW - my profile seems to indicate a different problem (but, I'm upgrading from 1.11):

```
# 2.5.6
| Top leaves:
| Babel.compile..........................................603,897 ms (25639)
| linker File#computeAssignedVariables...................488,136 ms (27564)
| other Target#minifyJs..................................474,948 ms (2)
| other ImportScanner#findImportedModuleIdentifiers.......43,497 ms (26261)
| other plugin templating-compiler........................38,965 ms (2)
```

```
# 1.11
| Top leaves:
| other Target#minifyJs..................................341,807 ms (2)
| files.readFile..........................................56,833 ms (207961)
| CssTools.minifyCss......................................42,789 ms (92)
| linker File#computeAssignedVariables....................33,872 ms (21044)
| Babel.compile...........................................32,736 ms (25780)
```

`computeAssignedVariables` is really interesting - for some modules it's spending 100+ seconds in them, it's wild. Not even always our biggest ones - e.g., it spent 70 seconds in `emojicons`.

**Comment by zodern** [comment]
@znewsham are rebuilds also slow for you? If so, could you please open a new issue for `computeAssignedVariables` being slow? It looks like we might need to adjust the cache size for that also. It would be helpful if you could follow the instructions in https://github.com/meteor/meteor/issues/11905#issuecomment-1042052218 to share the full profile for the initial build and a rebuild.

**Comment by znewsham** [comment]
@zodern sorry - was on vacation, I haven't tried dev (re)builds with a profile yet - I've been focused on prod builds, but I will take a look now.

@renanccastro are we blocked on a 2.5 version until zodern's question is resolved?

**Comment by znewsham** [comment]
@zodern initial dev build isn't unreasonable (about 5 minutes, it's a huge app) I've posted the trace (I set `METEOR_PROFILE=1`) and the same for the rebuild - it seems on the dev side, things are reasonably fast, it's just when making production builds they are intolerably slow
[dev-rebuild-2.5.txt](https://github.com/meteor/meteor/files/8134549/dev-rebuild-2.5.txt)
[dev-build-2.5.txt](https://github.com/meteor/meteor/files/8134550/dev-build-2.5.txt)

**Comment by Isarstyle** [comment]
I do not understand how a rebuild time in a dev environment of 30+ seconds is reasonably fast.. when there are options available outside meteor that do a "real" react-fast-refresh almost instantly.

**Comment by ouya99** [comment]
@Isarstyle issue is about build times not rebuild times. for HMR see https://github.com/meteor/meteor/issues/11936

**Comment by zodern** [comment]
@znewsham thanks for sharing the profiles. Could you please create a new issue for the build performance issues with your app?

**Comment by znewsham** [comment]
Yeah, can do - I was just going to post again, I tried cloning the `babel-compiler` package, and build times went even higher, this is with a pre-image cache, virtually none of the files should have required rebuild :( 

```
| Top leaves:
| linker File#computeAssignedVariables.................6,408,668 ms (27564)
| Babel.compile..........................................693,220 ms (25639)
| other Target#minifyJs..................................544,547 ms (2)
| other ImportScanner#findImportedModuleIdentifiers.......63,561 ms (26447)
| other plugin templating-compiler........................43,306 ms (2)
| getPrelinkedFiles toStringWithSourceMap.................41,690 ms (1198)

```

**Comment by znewsham** [comment]
As requested @zodern https://github.com/meteor/meteor/issues/11957

**Comment by zodern** [comment]
Thanks @znewsham. Your app's rebuilds are also much slower than we want. Would you mind opening an issue for slow rebuilds too? Our goal is to get them down to within a couple of seconds.

**Comment by znewsham** [comment]
For sure, I'll be very impressed and super happy if that was the case. Our
app is really large with lots of local packages - not saying there isn't
room for improvement though!

**Comment by znewsham** [comment]
@zodern here we go: https://github.com/meteor/meteor/issues/11958

**Comment by josherich** [comment]
babel/parser@7.16.12 has perf issue, which could be the main factor of slowness in `computeAssignedVariables` and `findImportedModuleIdentifiers`

according to https://github.com/babel/parser_performance
<img width="873" alt="image" src="https://user-images.githubusercontent.com/1488391/165395954-f609b133-0573-4740-afca-82fc05e2df07.png">
